drivers/adc/adc_emul.c

```diff
@@ -0,0 +1,512 @@
+/**
+ * @file
+ *
+ * @brief Emulated ADC driver
+ */
+
+/*
+ * Copyright 2021 Google LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT zephyr_adc_emul
+
+#include <drivers/adc.h>
+#include <drivers/adc/adc_emul.h>
+#include <kernel.h>
+#include <logging/log.h>
+#include <sys/byteorder.h>
+#include <sys/util.h>
+#include <zephyr.h>
+
+LOG_MODULE_REGISTER(adc_emul, CONFIG_ADC_LOG_LEVEL);
+
+#define ADC_CONTEXT_USES_KERNEL_TIMER
+#include "adc_context.h"
+
+#define ADC_EMUL_MAX_RESOLUTION 16
+
+typedef uint16_t adc_emul_res_t;
+
+struct adc_emul_channel_cfg {
+	adc_emul_value_function func;
+	void *param;
+	enum adc_gain gain;
+	enum adc_reference ref;
+};
+
+struct adc_emul_config {
+	uint8_t channels_num;
+};
+
+struct adc_emul_data {
+	struct adc_context ctx;
+	const struct device *dev;
+	uint16_t *buffer;
+	uint16_t *repeat_buffer;
+	uint32_t channels;
+	uint16_t res_mask;
+	uint16_t ref_vdd;
+	uint16_t ref_ext0;
+	uint16_t ref_ext1;
+	uint16_t ref_int;
+	uint8_t oversampling;
+	uint8_t differential;
+	struct adc_emul_channel_cfg *channels_cfg;
+	uint32_t *const_values;
+	struct k_thread thread;
+	struct k_sem sem;
+
+	K_KERNEL_STACK_MEMBER(stack,
+			CONFIG_ADC_EMUL_ACQUISITION_THREAD_STACK_SIZE);
+};
+
+/**
+ * @brief Standard ADC value function used when channel should return constant
+ *        value
+ *
+ * @param dev ADC emulator device
+ * @param channel Channel for which function is called
+ * @param param Pointer to mV value which should be returned
+ * @param result Pointer where result should be saved
+ *
+ * @return Always 0
+ */
+static int adc_emul_const_channel_value(const struct device *dev,
+					uint8_t channel, void *param,
+					uint32_t *result)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(channel);
+
+	*result = *(uint32_t *)param;
+
+	return 0;
+}
+
+int adc_emul_const_value_set_masked(const struct device *dev, uint32_t channels,
```

Comment:
Would it be good enough to just set it for a single channel? It would make the API and these functions simpler. The caller could add a loop if needed.

What is the benefit here?

Reply:
It is just more convenient to set multiple channels in one call. But you are right that it is easy enough to call single channel API in loop. I have removed `masked` variants of API.